### PR TITLE
code suggestions for unused-mut, while-true, deprecated-attribute, and unused-parens lints

### PR DESCRIPTION
**Pull request by zackmdavis**
lint_suggestions

r? @estebank

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -76,9 +76,13 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for WhileTrue {
             if let hir::ExprLit(ref lit) = cond.node {
                 if let ast::LitKind::Bool(true) = lit.node {
                     if lit.span.ctxt() == SyntaxContext::empty() {
-                        cx.span_lint(WHILE_TRUE,
-                                    e.span,
-                                    "denote infinite loops with loop { ... }");
+                        let msg = "denote infinite loops with `loop { ... }`";
+                        let mut err = cx.struct_span_lint(WHILE_TRUE,
+                                                          e.span, msg);
```
Comment (estebank) [inline src/librustc_lint/builtin.rs]: Fits in one line?

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -76,9 +76,13 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for WhileTrue {
...
+                        let condition_span = cx.tcx.sess.codemap().def_span(e.span);
+                        err.span_suggestion_short(condition_span, "use `loop`",
+                                                  "loop".to_owned());
```
Comment (estebank) [inline src/librustc_lint/builtin.rs]: When all arguments don't fit, I like to keep one line per argument (indented to the first one in the same line as the method).

File: src/librustc_lint/unused.rs [hunk 3]
```diff
@@ -325,9 +330,40 @@ impl UnusedParens {
...
                             parser::contains_exterior_struct_lit(&inner);
             if !necessary {
-                cx.span_lint(UNUSED_PARENS,
-                             value.span,
-                             &format!("unnecessary parentheses around {}", msg))
+                let span_msg = format!("unnecessary parentheses around {}", msg);
+                let mut err = cx.struct_span_lint(UNUSED_PARENS,
+                                                  value.span,
+                                                  &span_msg);
+                // Remove exactly one pair of parentheses (rather than naïvely
```
Comment (estebank) [inline src/librustc_lint/unused.rs]: @nikomatsakis do we care about unicode in rustc comments?

**Review by estebank — APPROVED** [review]
Small nitpicks, r=me once you deal with the first two.

**Comment by zackmdavis** [comment]
(force-pushed to fix formatting)

**Comment by zackmdavis** [comment]
(updated again to fix compile-fail failures reported by Travis)

**Comment by estebank** [comment]
@bors r+

Thanks @zachreizner, great work!

**Comment by zackmdavis** [comment]
> Thanks @zachreizner, great work!

I agree; @zachreizner is the best! I especially like [the one where he fixed the shift-right docs](https://github.com/rust-lang/rust/pull/30370/files); it must have taken a very careful eye to catch that!

:wink:

**Comment by zachreizner** [comment]
Not the first time I've been thanked in place of a a fellow Zach :-). I think my handle just comes up before the Zac**k** Zach-s in the autocomplete.

**Comment by zackmdavis** [comment]
Appveyor/Windows apparently [doesn't like](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4847/job/6unpnf25oepts2q2) the `--error-format json` UI test. I think this is _plausibly_ a bug in the test runner (rather than the JSON order being unstable or anything like that), since the only difference seems to be in the `"file_name"` value, which is supposed to be normalized (with "`$DIR`") and worked fine in the Travis/Unix-like tests, but for now I'll just rewrite it as a run-make test.

**Comment by estebank** [comment]
I believe we're _really_ trying to move away from `make` tests. @zackmdavis, could you

1) create a ticket to deal with the test output normalization failure on json output and
2) revert your last change and add `// ignore-windows` with a comment linking to the previous ticket?

**Comment by zackmdavis** [comment]
output normalization issue is #44968; updated tip of PR with ignore-windows is 979671ec

**Comment by zackmdavis** [comment]
**correction**, PR is 8a14022c

**Pull request merged**